hmetrics/hmetrics.go

```diff
@@ -44,61 +50,84 @@ func (e HerokuMetricsURLUnset) Fatal() bool {
 	return true
 }
 
+// ErrHandler funcations are used to provide custom processing/handling of
+// errors encountered during the collection or reporting of metrics to Heroku.
+type ErrHandler func(err error) error
+
 var (
 	mu      sync.Mutex
 	started bool
 )
 
-// ErrHandler receives any errors encountered during collection or reporting of metrics to Heroku. Processing of metrics
-// continues if the ErrHandler returns nil, but aborts if the ErrHandler itself returns an error.
-type ErrHandler func(err error) error
-
+// Report go metrics to Heroku until the context is canceled.
+//
+// Only one call to the ErrHandler will happen at a time. Metrics can be dropped
+// or delayed if a call to ErrHandler takes longer than the reprting interval.
+// Processing of metrics continues if the ErrHandler returns nil, but aborts if
+// the ErrHandler itself returns an error. It is safe to pass a nil ErrHandler.
+//
+// Report is safe for concurrent usage, but calling it more than once w/o
+// canceling the context returns an AlreadyStarted error. This is to ensure that
+// metrics aren't duplicated. Report can be called again to restart reporting
+// after the context is canceled.
 func Report(ctx context.Context, ef ErrHandler) error {
 	mu.Lock()
-	defer mu.Unlock()
 	if started {
```

Comment:
I would break the code before and after report into two functions to make it clearer what's happening here.  As it is it's kind of confusing, maybe the following?  Also, I'm thinking that the public DefaultEndpoint is going to be a problem for concurrent use.
```
func Report(ctx context.Context, ef ErrHandler) error 
       if endpoint, err := alreadyStarted(); err != nil {
            return err 
       }
	report(ctx, &http.Client{Timeout: 20 * time.Second}, endpoint, errorHandler(ef))
        setStarted() 
	return nil
}

func alreadyStarted() string, error {
      mu.Lock() 
      defer  func() {
          started = localStarted
          mu.Unlock() 
     }() 
     var localStarted bool
      
	if started {
		return "", AlreadyStarted{}
	}
        
	if DefaultEndpoint == "" {
		return "" HerokuMetricsURLUnset{}
	}
	localStarted = true
        endpoint := DefaultEndpoint
        return endpoint,  nil 
}
```

Reply:
@murphybytes I did something similar, but different. Please take a look and if you are okay, approve. Thx.